### PR TITLE
Support subsampling in AutoDelta, AutoNormal

### LINKED ISSUE #2322: [FR] Subsampling in AutoNormal and AutoDelta
This issue discusses interface design for subsampling in `AutoGuide`s.

## Motivation

I'd like to support subsampling in `pyro.contrib.forecast` #2311 , e.g. for the [upcoming BART tutorial](https://nbviewer.jupyter.org/github/pyro-ppl/pyro/blob/forecast-tutorial/tutorial/source/forecasting.ipynb).

## Background

After #2320, `AutoNormal` will know about each site's `event_dim`. With only a little more work we could make `AutoNormal` subsample-compatible. It would be easy to make similar changes to `AutoDelta`.

`EasyGuide` already supports subsampling. The interface requires users to subclass `EasyGuide`, define a method `.guide(self, *args, *kwargs)`, and within that method call `self.plate("name", size, **subsample_kwargs)`. For example: https://github.com/pyro-ppl/pyro/blob/ac8b7e72916478783a39482dab4dbf71a9325127/tests/contrib/easyguide/test_easyguide.py#L79-L80

## Proposals for possible interfaces

1. Update `AutoGuide.forward()` to accept `subsample` and/or `subsample_size` dicts mapping name to subsample index tensor subsample size, resp.
  **Pros:** Straight-forward to implement
  **Cons:** Unclear how to interface with ELBO, SVI
2. Allow users to override a method `.create_plates(self, *args, **kwargs)`.
  **Pros:** Clear how to integrate with ELBO,  SVI
  **Cons:** Verbose, clumsy
3. Update `AutoGuide.__init__()` to accept a `create_plates(*args, **kwargs)` function that creates zero or more plates from the model's `*args, **kwargs`, and returns them.
  **Pros:** Clear how to integrate with ELBO, SVI; clear how to integrate with `Forecaster`
  **Cons:** ???

### PR DESCRIPTION
**Pull request by fritzo**
Resolves #2322
Addresses #2311 

This supports subsampling by adding a `create_plates()` function to `AutoDelta.__init__()` and `AutoNormal.__init__()`. For example with origin-destination-time data:
```py
def model(odt_data, subsample_origins=None, subsample_destins=None):
    # ...user defined...

def create_plates(odt_data, subsample_origins=None, subsample_destins=None):
    O, D, T = odt_data.shape
    return [pyro.plate("origin", size=O, subsample_size=subsample_origins, dim=-3),
            pyro.plate("destin", size=D, subsample_size=subsample_destins, dim=-2)]

guide = AutoNormal(model, create_plates=create_plates)
```

This also includes the following fixes:
-  introduces an experimental `pyro.subsample()` primitive (pair coded with @eb8680)
- changes `dist.Uniform` to support unbroadcasted params for broadcastable `.support` shape

## Tested
- [x] `dist.Uniform` refactoring is covered by existing tests
- [x] tests for `pyro.subsample()`
- [x] `AutoDelta` tests
- [x] `AutoNormal` tests
- [x] forecaster tests

### CODE REVIEW

File: pyro/infer/autoguide/guides.py [hunk 1]
```diff
@@ -306,18 +321,35 @@ def my_init_fn(site):
...
+            value = site["value"].detach()
+            event_dim = site["fn"].event_dim
+
+            # If subsampling, repeat init_value to full size.
+            for frame in site["cond_indep_stack"]:
+                plate = self._plates[frame.name]
+                if plate.subsample_size < plate.size:
+                    dim = plate.dim - event_dim
+                    assert value.size(dim) == plate.subsample_size
+                    value = periodic_repeat(value, plate.size, dim).contiguous()
```
Comment (fritzo) [inline pyro/infer/autoguide/guides.py]: This logic was adapted from `EasyGuide.map_estimate()` which already supports subsampling.
https://github.com/pyro-ppl/pyro/blob/ac8b7e72916478783a39482dab4dbf71a9325127/pyro/contrib/easyguide/easyguide.py#L148-L153

File: pyro/infer/autoguide/guides.py [hunk 2]
```diff
@@ -306,18 +323,33 @@ def my_init_fn(site):
...
-            value = PyroParam(site["value"].detach(), constraint=site["fn"].support)
+            value = site["value"].detach()
+            event_dim = site["fn"].event_dim
+
+            # If subsampling, repeat init_value to full size.
+            for frame in site["cond_indep_stack"]:
+                full_size = getattr(frame, "full_size", frame.size)
+                if full_size != frame.size:
+                    dim = frame.dim - event_dim
+                    value = periodic_repeat(value, full_size, dim).contiguous()
```
Comment (fritzo) [inline pyro/infer/autoguide/guides.py]: This logic was adapted from `EasyGuide.map_estimate()` which already supports subsampling.
https://github.com/pyro-ppl/pyro/blob/ac8b7e72916478783a39482dab4dbf71a9325127/pyro/contrib/easyguide/easyguide.py#L143-L153

Comment (martinjankowiak) [inline pyro/infer/autoguide/guides.py]: why periodic? for more variability?

Comment (fritzo) [reply pyro/infer/autoguide/guides.py]: Yes, and also in the special case of sequential window minibatching this preserves the first minibatch. This is the same strategy used in `EasyGuide`.

File: pyro/infer/autoguide/guides.py [hunk 3]
```diff
@@ -394,12 +439,19 @@ def _setup_prototype(self, *args, **kwargs):
         # Initialize guide params
         for name, site in self.prototype_trace.iter_stochastic_nodes():
             # Collect unconstrained event_dims, which may differ from constrained event_dims.
-            init_loc = biject_to(site["fn"].support).inv(site["value"].detach())
+            init_loc = biject_to(site["fn"].support).inv(site["value"].detach()).detach()
```
Comment (martinjankowiak) [inline pyro/infer/autoguide/guides.py]: why double `detach()`? wrap in`no_grad()` instead?

File: tests/contrib/forecast/test_forecaster.py [hunk 4]
```diff
@@ -118,3 +118,83 @@ def test_smoke(Model, batch_shape, t_obs, t_forecast, obs_dim, cov_dim, dct_grad
...
+            duration, obs_dim = zero_data.shape[-2:]
+            noise_dist = dist.GaussianHMM(
+                dist.Normal(0, 1).expand([obs_dim]).to_event(1),
+                torch.eye(obs_dim),
+                dist.Normal(0, 1).expand([obs_dim]).to_event(1),
+                torch.eye(obs_dim),
+                dist.Normal(0, 1).expand([obs_dim]).to_event(1),
+                duration=duration,
+            )
+            # FIXME Data and covariates need to be subsampled.
```
Comment (martinjankowiak) [inline tests/contrib/forecast/test_forecaster.py]: stale comment?

File: tests/contrib/forecast/test_forecaster.py [hunk 5]
```diff
@@ -118,3 +118,83 @@ def test_smoke(Model, batch_shape, t_obs, t_forecast, obs_dim, cov_dim, dct_grad
...
+            covariates = pyro.subsample(covariates, event_dim=1)
+
+            loc = zero_data[..., :1, :]
+            scale = pyro.sample("scale", dist.LogNormal(loc, 1).to_event(1))
+
+            with self.time_plate:
+                jumps = pyro.sample("jumps", dist.Normal(0, scale).to_event(1))
+            prediction = jumps.cumsum(-2)
+
+            # FIXME Expansion of constraints.interval breaks guide shapes.
```
Comment (martinjankowiak) [inline tests/contrib/forecast/test_forecaster.py]: stale comment?

File: pyro/primitives.py [hunk 6]
```diff
@@ -148,6 +148,40 @@ def deterministic(name, value, event_dim=None):
     return sample(name, dist.Delta(value, event_dim=event_dim).mask(False), obs=value)
 
 
+@effectful(type="subsample")
+def subsample(data, event_dim):
+    """
+    EXPERIMENTAL Subsampling statement to subsample data based on enclosing
```
Comment (eb8680) [inline pyro/primitives.py]: nit: make this a Sphinx warning?

File: pyro/poutine/subsample_messenger.py [hunk 7]
```diff
@@ -132,7 +133,7 @@ def _process_message(self, msg):
         msg["scale"] = msg["scale"] * self.size / self.subsample_size
 
     def _postprocess_message(self, msg):
-        if msg["type"] == "param" and self.dim is not None:
+        if msg["type"] in ("param", "subsample") and self.dim is not None:
```
Comment (eb8680) [inline pyro/poutine/subsample_messenger.py]: The error message in 145-147 will be incorrect for `subsample`s because they have no `msg["name"]`

File: pyro/infer/autoguide/guides.py (hunk 3 above)
Comment (fritzo) [reply pyro/infer/autoguide/guides.py]: Even if we wrapped in `no_grad()` we'd need to detach, since `biject_to()` might be simply `lambda x: x`.

File: pyro/primitives.py (hunk 6 above)
Comment (fritzo) [reply pyro/primitives.py]: This is how we mark experimental everywhere. I can change that convention, but we should do so in a separate PR that changes all relevant docstrings.

File: pyro/poutine/subsample_messenger.py (hunk 7 above)
Comment (fritzo) [reply pyro/poutine/subsample_messenger.py]: fixed.

File: tests/contrib/forecast/test_forecaster.py (hunk 5 above)
Comment (fritzo) [reply tests/contrib/forecast/test_forecaster.py]: removed.

File: tests/contrib/forecast/test_forecaster.py (hunk 4 above)
Comment (fritzo) [reply tests/contrib/forecast/test_forecaster.py]: removed.
